### PR TITLE
[packaging] Initial packaging of cbindgen 0.17.0. Fixes JB#52894

### PR DESCRIPTION
**Pull request by paveltkv**

### CODE REVIEW

File: rpm/rust-cbindgen.spec [hunk 1]
```diff
@@ -0,0 +1,41 @@
+%global rustflags -Clink-arg=-Wl,-z,relro,-z,now
+
+Name:           cbindgen
+Version:        0.16.0
+Release:        0
+Summary:        A tool for generating C bindings from Rust code
+License:        MPL-2.0
```
Comment (rainemak) [inline rpm/rust-cbindgen.spec]:
```suggestion
License:        MPLv2.0
```
See https://fedoraproject.org/wiki/Licensing:Main#Good_Licenses

File: rpm/rust-cbindgen.spec [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+Source:         %{name}-%{version}.tar.bz2
+BuildRequires:  cargo >= 1.30.0
+BuildRequires:  rust >= 1.30.0
+BuildRequires:  rust-std-static >= 1.30.0
+
+%description
+A tool for generating C bindings from Rust code.
+
+%prep
+%autosetup -n %{name}-%{version_no_tilde}/cbindgen -p1
```
Comment (okodron) [inline rpm/rust-cbindgen.spec]: Minor cosmetic proposal:
```suggestion
%autosetup -p1 -n %{name}-%{version}/%{name}
```

File: rpm/rust-cbindgen.spec [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+# will try to download source deps and rebuild
+export CARGO_HOME=`pwd`/cargo-home/
+# cargo install appends /bin to the path
+cargo install --root=%{buildroot}%{_prefix} --path .
+# remove spurious files
+rm -f %{buildroot}%{_prefix}/.crates.toml
+rm -f %{buildroot}%{_prefix}/.crates2.json
+
+%files
+%license LICENSE
```
Comment (rainemak) [inline rpm/rust-cbindgen.spec]: %doc contributing.md docs.md internals.md CHANGES README.md

As this is only a tool for building.

File: rpm/rust-cbindgen.spec [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  -exec sh -c 'echo "{\"files\":{ },\"package\":\"\"}" > "$(dirname $0)/.cargo-checksum.json"' '{}' \;
+
+# Remove dependency checksums
+sed -i 's/checksum = "[^"]*"/checksum = ""/' Cargo.lock
+
+%build
+export RUSTFLAGS="%{rustflags}"
+export CARGO_HOME=`pwd`/cargo-home/
+
+cargo build --frozen --release
```
Comment (rainemak) [inline rpm/rust-cbindgen.spec]:
```suggestion
cargo build --offline --frozen --release
```
Adding offline would document nicely also from build point of view that this doesn't require networking.

File: .gitmodules [hunk 5]
```diff
@@ -0,0 +1,123 @@
+[submodule "cbindgen"]
+	path = cbindgen
+	url = https://github.com/eqrion/cbindgen
```
Comment (rainemak) [inline .gitmodules]: @mkosola I think we need to create still mirrors for each of these. Could you help with those?

**Review by rainemak — COMMENTED** [review]
Adding review comments. Looking very nice already.

File: rpm/rust-cbindgen.spec (hunk 1 above)
Comment (rainemak) [reply rpm/rust-cbindgen.spec]:
```suggestion
License:        MPLv2.0 and MIT and BSD and (ASL 2.0 or Boost)
```

Actually source package could have only this MPLv2.0 but for the compiled cbindgen we need cover also submodules.

File: rpm/rust-cbindgen.spec [hunk 6]
```diff
@@ -0,0 +1,70 @@
+%global rustflags -Clink-arg=-Wl,-z,relro,-z,now
+
```
Comment (rainemak) [inline rpm/rust-cbindgen.spec:3]: %global crate cbindgen

File: rpm/rust-cbindgen.spec [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+Name:           cbindgen
```
Comment (rainemak) [inline rpm/rust-cbindgen.spec]:
```suggestion
Name:           rust-%{crate}
```

File: rpm/rust-cbindgen.spec [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+Version:        0.16.0
+Release:        0
+Summary:        A tool for generating C bindings from Rust code
+License:        MPL-2.0
```
Comment (rainemak) [inline rpm/rust-cbindgen.spec]: Changing this to rust-cbindgen would change this to MPLv2.0 only.

File: rpm/rust-cbindgen.spec [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+URL:            https://crates.io/crates/cbindgen
+Source:         %{name}-%{version}.tar.bz2
+BuildRequires:  cargo >= 1.30.0
+BuildRequires:  rust >= 1.30.0
+BuildRequires:  rust-std-static >= 1.30.0
+
+%description
+A tool for generating C bindings from Rust code.
+
```
Comment (rainemak) [inline rpm/rust-cbindgen.spec:29]: Then we could introduce the package for the cbindgen as...

```
%package     -n %{crate}
Summary:        %{summary}
License:        MPLv2.0 and MIT and BSD and (ASL 2.0 or Boost)
```

Comment (rainemak) [reply rpm/rust-cbindgen.spec:29]: See https://src.fedoraproject.org/rpms/rust-cbindgen/blob/master/f/rust-cbindgen.spec

**Review by rainemak — COMMENTED** [review]
Let's keep the rust-cbindgen package as the source package have cbindgen as a subpackage.

File: .gitmodules (hunk 5 above)
Comment (rainemak) [reply .gitmodules]: I think each rust crate mirror should be prefixed with "rust-" similar to this repository. The pattern would then be the same as used by Fedora and it's good to align with some other distro.

Comment (mkosola) [reply .gitmodules]: https://git.sailfishos.org/mirror/rust-cbindgen
https://git.sailfishos.org/mirror/rust-ansi-term
https://git.sailfishos.org/mirror/rust-atty
https://git.sailfishos.org/mirror/rust-autocfg
https://git.sailfishos.org/mirror/rust-bitflags
https://git.sailfishos.org/mirror/rust-cfg-if
https://git.sailfishos.org/mirror/rust-clap
https://git.sailfishos.org/mirror/rust-cloudabi
https://git.sailfishos.org/mirror/rust-getrandom
https://git.sailfishos.org/mirror/rust-hashbrown
https://git.sailfishos.org/mirror/rust-heck
https://git.sailfishos.org/mirror/rust-rusty-hermit
https://git.sailfishos.org/mirror/rust-indexmap
https://git.sailfishos.org/mirror/rust-itoa
https://git.sailfishos.org/mirror/rust-lazy-static.rs
https://git.sailfishos.org/mirror/rust-libc
https://git.sailfishos.org/mirror/rust-parking_lot
https://git.sailfishos.org/mirror/rust-log
https://git.sailfishos.org/mirror/rust-cryptocorrosion
https://git.sailfishos.org/mirror/rust-proc-macro2
https://git.sailfishos.org/mirror/rust-quote
https://git.sailfishos.org/mirror/rust-rand
https://git.sailfishos.org/mirror/rust-syscall
https://git.sailfishos.org/mirror/rust-remove_dir_all
https://git.sailfishos.org/mirror/rust-ryu
https://git.sailfishos.org/mirror/rust-scopeguard
https://git.sailfishos.org/mirror/rust-serde
https://git.sailfishos.org/mirror/rust-json
https://git.sailfishos.org/mirror/rust-serial_test
https://git.sailfishos.org/mirror/rust-smallvec
https://git.sailfishos.org/mirror/rust-strsim-rs
https://git.sailfishos.org/mirror/rust-syn
https://git.sailfishos.org/mirror/rust-tempfile
https://git.sailfishos.org/mirror/rust-textwrap
https://git.sailfishos.org/mirror/rust-toml-rs
https://git.sailfishos.org/mirror/rust-unicode-segmentation
https://git.sailfishos.org/mirror/rust-unicode-width
https://git.sailfishos.org/mirror/rust-unicode-xid
https://git.sailfishos.org/mirror/rust-vec-map
https://git.sailfishos.org/mirror/rust-wasi
https://git.sailfishos.org/mirror/rust-winapi-rs

**Review by rainemak — APPROVED** [review]
LGTM

**Pull request merged**